added Smithsonian API service

dragazo opened the pull request:
This is Devin Jean; pardon the github username, I can't really change it at this point without breaking a lot of references.
This pull request currently includes a functioning Smithsonian API service, but doesn't yet have the public interface test - I'll be adding that soon, but this is the important part.

dragazo commented on src/server/services/procedures/smithsonian/smithsonian.js:
```diff
@@ -0,0 +1,118 @@
+ * @param {String} query JSON query string to use
+ */
+Smithsonian._contentQuery = function(id, query) {
+    return this._content(id).then(res => this.__queryJson(res, query)).catch(err => this.response.status(500).send(err));
```
Ah, ok. I got that pattern from other services in the repo, so in that case we could clean that up elsewhere as well.

brollb replied:
Yeah, that makes sense. Unfortunately, some of the other ones still need to be updated but there always seem to be more pressing things (of course, PRs are welcome :) ). Have you seen the [best practices doc](https://github.com/NetsBlox/NetsBlox/wiki/Best-Practices-for-NetsBlox-Services)?

dragazo replied:
I found that best practices page earlier today, which I probably should have looked for before starting. That covers some of the points you made here. I think it should conform after I make these smaller changes. By bigger concern is with the unfortunate behavior of the filtering with respect to the `count` parameter.

dragazo commented:
@brollb Yeah, I originally wasn't filtering search results at all, but then that meant that most (85-97.5%) of search results didn't have images. So the second param (max matches) is the fetch count, which is then manually filtered because their API doesn't have a way to do that on the search itself. This results in the matches count needing to be artificially high to get results (I was using 500 for the brief demo I did on zoom). But I think that's an unfortunate limitation of their search API.

Also, about the extra content from the raw API result: some of those would definitely be useful, but they might be inconsistently present for any given object like image data is. If we want all the exposed fields to be present this could mean even more filtering, otherwise we could return it all in some cleaned-up form, but that might just offset the headache to the user if field didn't have values.

Oh, and one other thing, I noticed you passed 1 as the first argument to the search. At the time of writing the code, I had forgotten scratch effectively 1-bases arrays, so I should probably rewrite that to be a 1-based index instead of 0..

brollb commented on src/server/services/procedures/smithsonian/smithsonian.js:
```diff
@@ -0,0 +1,118 @@
+ * Searches the EDAN db and returns up to count ids denoting matching historical objects
+ * 
+ * @param {String} term Term to search for
+ * @param {BoundedNumber<0>} start Starting index of search (default 0)
```
Makes sense. Optional arguments are supported by NetsBlox RPCs so that isn't an issue.

Yeah, it isn't obvious what the best approach is for these sorts of parameters. Do you know how many results are returned generally? I lean toward keeping them but maybe increasing the limit to something like 100.

brollb replied:
Also, the limit should correspond to the number of results returned to the user - not the limit before filtering the results.

dragazo replied:
Yeah, I agree on the returned to user vs before filter point, but as far as I can see there's not really a way to do that neatly given that it would require an arbitrary number of API calls due to them not supporting it at the API level. Your point about having both filter and non-filter versions could solve that, though. As long as it's not confusing at a user level that images aren't guaranteed to exist unless using the filtering search.

brollb commented:
Good point about 1 indexing. I actually started with `0` but then replaced it with `1` just in case. (This was before I read the code.) Another way around it would be to rename the argument to `skip`.

I wonder if it would be good to have RPCs for both searching generally and searching for data with images. If there is any interest in the data without images, I would lean toward this approach. Unfortunately, I am not sure how much interest there is for this type of info. Maybe it would be interesting to look at the distribution of results for various regions or time periods?

dragazo commented:
Addressing the new commit:
Minor: Removed the `.catch` patterns. Changed the Smithsonian API Key to `Data.gov` with the environment name `DATA_GOV_KEY`. Changed the `start` param for search to `skip` and left as 0-based.
Major: In light of the fact that the raw API call `search` returns a lot of the info we just refer to later, I've decided to take the list of lists approach and package it all into just the `search` call. This pattern came from implementing a new, very similar, but separate service: New York Public Library (NYPL). It's only included in this PR because they're basically the same thing but from different sources. NYPL seems to have better/more content, but for some reason the image fetching is super slow (seems to be a latency problem, not download speed, so maybe high traffic?).

brollb commented on src/server/services/procedures/smithsonian/smithsonian.js:
```diff
@@ -7,112 +7,87 @@
+
 /**
- * Gets a thumbnail associated with the given object
+ * Gets an image associated with the given object
```
"Gets" -> "Get"

brollb commented on src/server/services/procedures/smithsonian/smithsonian.js:
```diff
@@ -7,112 +7,87 @@
-    return this.thumbnailUrls(id)
-        .then(urls => this._sendImage({url:urls[0]}))
-        .catch(err => this.response.status(500).send(err));
+Smithsonian.getImage = function(entry) {
```
`entry` -> `ID`

brollb commented on src/server/services/procedures/new-york-public-library/new-york-public-library.js:
```diff
@@ -0,0 +1,106 @@
+            }
+        }
+
+        return [dateIssued, place, publisher, genres, subjects];
```
This should return an object as it isn't clear to the user what these values are (when calling the RPC from NetsBlox). Something like:
```javascript
return {
  'date issued': dateIssued,
  place,
  publisher,
  genres,
  subjects,
};
```

brollb commented on src/server/services/procedures/new-york-public-library/new-york-public-library.js:
```diff
@@ -0,0 +1,106 @@
+/**
+ * Gets the image links associated with the given entry returned from search().
+ * 
+ * @param {Array} entry An entry returned from search()
```
Or just `ID` might be even better to match the entry.

brollb commented on src/server/services/procedures/new-york-public-library/new-york-public-library.js:
```diff
@@ -0,0 +1,106 @@
+ * 
+ * @param {Array} entry An entry returned from search()
+ */
+NYPL.getImage = function(entry) {
```
Actually, `imageID` would probably be better.

brollb commented on src/server/services/procedures/new-york-public-library/new-york-public-library.js:
```diff
@@ -0,0 +1,106 @@
+        for (const item of listify(results)) {
+            const itemURL = item.apiItemURL;
+            const itemID = itemURL.substr(itemURL.lastIndexOf('/') + 1);
+            ret.push([item.uuid, item.title, item.dateDigitized, itemID, item.imageID]);
```
Actually, if you use `imageID` as the argument for `getImage`, it might be nice to use camelcase here, too:
```javascript
ret.push({
  ID: itemID,
  title: item.title,
  dateDigitized: item.dateDigitized,
  imageID: item.imageID,
});
```

dragazo replied:
`itemID` can be used to generate the image URL directly without going through another API call, which is what the code used to do. Unfortunately, it also requires an image format type, which is not guaranteed to exist, so in the end I just rewrote it to do the API call anyway. So it's not used anymore - I'll remove it for clarity.

dragazo replied:
I think `uuid` and `itemID` are artifacts of some older version. The `itemID` is used for the `/items/` API call, while the `uuid` is used for everything else, like `/items/mods/`.

brollb commented on src/server/services/procedures/new-york-public-library/new-york-public-library.js:
```diff
@@ -0,0 +1,106 @@
+        path:`items/mods/${entry[0]}`,
+        headers:{Authorization:`Token token=${this.apiKey.value}`},
+    }).then(res => {
+        const mods = res.nyplAPI.response.mods;
```
Need to handle the case where nothing is found (such as an invalid ID being used). Currently, it responds with `Cannot read property 'originInfo' of undefined`

dragazo replied:
Ok, I've changed `details` and `imageURLs` to return empty on invalid input.

brollb commented on src/server/services/procedures/new-york-public-library/new-york-public-library.js:
```diff
@@ -0,0 +1,106 @@
+ * @param {Array} entry An entry returned from search()
+ */
+NYPL.getImage = function(entry) {
+    return this.imageURLs(entry).then(urls => this._sendImage({url:urls[0]}));
```
Need to handle the case where no image is found. Current error is `Cannot read property '0' of undefined`

brollb replied:
It also might be good to select the image more intelligently since the different links can vary a lot in size (http://api.repo.nypl.org/#image-links). I would probably look for either 760 px (w) or 300 px (r) versions. If those don't exist, then whatever one is first...

brollb commented:
Using lists of lists and packaging it all in the search call sounds great. I should've read your comment in more detail before reviewing the code :) 
I wouldn't worry too much about the slow API. It seems fast enough and should be faster after the first request since we are caching requests on our side.

dragazo commented on src/server/services/procedures/new-york-public-library/new-york-public-library.js:
```diff
@@ -0,0 +1,106 @@
+ * Searches the NYPL db an returns up to count ids denoting matching items.
+ * 
+ * @param {String} term Term to search for
+ * @param {BoundedNumber<1>} per_page Maximum number of items to return for any given page
```
Ah, right, I've been in rust land too long - I'll change to cammelcase and give default.

dragazo replied:
Ok, I added a default and updated jsdoc with the `=` thing ({`BoundedNumber<1>=}` and it seems to work in NetsBlox, but it doesn't show any indication of what the default value is - just looks empty. Is that expected?

brollb replied:
Yeah, unfortunately that is the case now and it needs to be added to the argument info manually. It would probably be good to support the JSDoc version listed [here](https://jsdoc.app/tags-param.html) rather than the Google Closure Compiler version (or maybe along with that version).

dragazo commented on src/server/services/procedures/new-york-public-library/new-york-public-library.js:
```diff
@@ -0,0 +1,106 @@
+        headers:{Authorization:`Token token=${this.apiKey.value}`},
+    }).then(res => {
+        const results = res.nyplAPI.response.result;
+        if (results === undefined) return [];
```
Right. `listify` didn't handle undefined when this was written. Removing it now.

dragazo commented:
I just rewrote the wrapping of search matches to be objects, but it looks like that's translated into NetsBlox as an nx2 array of key-value pairs. I'm a little concerned that the order could be undefined due to being essentially a hashmap converted to a list. Is there a way to index it with a key? Otherwise getting e.g. the uuid could be annoying since it might not be item 1. Or I might be wrong about all this and it'll be in right order always?

brollb commented:
Yeah, that's a good question. The key order should follow chronological order: https://www.stefanjudis.com/today-i-learned/property-order-is-predictable-in-javascript-objects-since-es2015/

dragazo commented:
I think the latest commit addresses all the areas you mentioned. For the Smithsonian service there's now a filtered and non-filtered version of search. For both Smithsonian and NYPL, if you try to get an image for something that doesn't have one it returns empty string (NetsBlox didn't seem to like `null` or `undefined`). Additionally, I fixed all the areas where giving an invalid uuid/id/itemid would show an error - most of these places returned arrays, so now just return empty array on failure. Contextually, this seems intuitive (e.g. `getImageURLs` on invalid id returns `[]` because technically a non-existent item doesn't have any).

brollb commented on src/server/services/procedures/smithsonian/smithsonian.js:
```diff
@@ -0,0 +1,113 @@
+ * @param {BoundedNumber<0>=} skip Number of items to skip from beginning
+ * @returns {Array} Up to count matches, each being [id, title, types[], authors[], topics[], notes[], physicalDescriptions[], sources[]]
+ */
+Smithsonian.searchUnfiltered = function(term, count = 100, skip = 0) {
```
I would maybe rename `searchUnfiltered`, `searchFiltered` to be `search` and `searchImageContent` so it is obvious what is being filtered out

brollb commented on src/server/services/procedures/smithsonian/smithsonian.js:
```diff
@@ -0,0 +1,113 @@
+ * 
+ * @param {String} id ID of an object returned from search
+ */
+Smithsonian.imageURLs = function(id) {
```
`imageURLs` -> `getImageURLs`

brollb commented on src/server/services/procedures/smithsonian/smithsonian.js:
```diff
@@ -0,0 +1,113 @@
+};
+
+/**
+ * Gets the image urls (0 or more) associated with the given object
```
`Gets` -> `Get`

brollb commented on src/server/services/procedures/new-york-public-library/new-york-public-library.js:
```diff
@@ -0,0 +1,109 @@
+ * @param {String} itemID itemID of the object
+ * @returns {Array} An array of imgurls[]
+ */
+NYPL.imageURLs = function(itemID) {
```
`imageURLs` -> `getImageURLs`

brollb commented on src/server/services/procedures/new-york-public-library/new-york-public-library.js:
```diff
@@ -0,0 +1,109 @@
+ * @param {String} uuid uuid of the object
+ * @returns {Array} Item details
+ */
+NYPL.details = function(uuid) {
```
`details` -> `getDetails`

brollb commented on src/server/services/procedures/new-york-public-library/new-york-public-library.js:
```diff
@@ -0,0 +1,109 @@
+}
+
+/**
+ * Search the New York Public Library collection.
```
I would add information about paging here. Maybe adding another sentence like: "Search results are stored in pages and the results on the designated page will be returned."

brollb replied:
This was one of the comments I was referring to before.

brollb reviewed (changes requested):
This is coming along nicely! There are a couple comments about the names of RPCs (for consistency), help message verb tense.

Also, tests for the RPC interfaces need to be added: https://github.com/NetsBlox/NetsBlox/blob/master/test/unit/server/services/procedures/geolocation.spec.js#L8-L18

dragazo commented:
New commit fixes the naming and comment conventions. I decided to use `searchForImages` for the filtered version of smithsonian search since I thought it would be a bit more clear for beginners (since showing images will probably be the first thing they'd want to do with the service, anyway).

brollb commented on src/server/services/procedures/smithsonian/smithsonian.js:
```diff
@@ -0,0 +1,113 @@
+ * @returns {Array} Up to count matches, each being [id, title, types[], authors[], topics[], notes[], physicalDescriptions[], sources[]]
+ */
+Smithsonian.searchForImages = function(term, count = 100, skip = 0) {
+    return this._raw_search(term, count, skip, item => {
```
It would probably be good to update this to actually fetch `count` results after filtering. One way to do this would be to use `async`/`await` and just perform raw_search in a loop until you have either run out of results or found `count` items.

brollb reviewed (changes requested):
Looking good. There are still 2 unresolved comments and it needs some tests (at least for the RPC interface to make sure we don't break backwards compatibility in the future). Then it should be good to merge into master and show up on dev.netsblox.org!

dragazo commented:
I changed filtered search to respect `count` and call in a loop. In my limited testing this makes around 2-20 API calls, and can therefore be somewhat slow. It currently just grabs `count` over and over until it has enough, but I could change it to be smarter about how to grab the data (e.g. extrapolate filtered density of first pass to predict total number needed before filtering), but then it risks grabbing way too much if the first pass had artificially low density.

One odd thing about the filtering: if the user wanted to skip 1000 filtered items, then grab 50 we would have to (essentially) grab 1050 after filtering. So I've decided to make the skip parameter always be prior to filtering. The search functions now return a pair: the matches array and the next skip value to use if the next page of results is needed.

Also each match has a new bool field called `hasImage` denoting if it had image data. I figure this would be useful if people wanted to use the raw, unfiltered search.

Also, I added the spec tests and ran them with `npm test` - everything passed except `geolocation > geolocate`, but I assume that's due to a missing API key on my local.

You mentioned there was another unresolved comment? The only other comment I could see was about the visibility of `getImageURLs`: it could be made private, but I figured there wasn't any harm in letting the users see them. So yeah, I could just hide it if you'd prefer.

brollb commented on src/server/services/procedures/smithsonian/smithsonian.js:
```diff
@@ -0,0 +1,125 @@
+ * Search and return up to count matching items
+ * 
+ * @param {String} term Term to search for
+ * @param {BoundedNumber<0>=} count Maximum number of items to return
```
`count` should probably be at least 1 rather than 0

brollb commented on src/server/services/procedures/smithsonian/smithsonian.js:
```diff
@@ -0,0 +1,125 @@
+ * Search and return up to count matching items (only ones with images)
+ * 
+ * @param {String} term Term to search for
+ * @param {BoundedNumber<0>=} count Maximum number of items to return
```
Same as before. Should be at least 1

brollb commented:
I agree about `getImageURLs` though I forget what the other comments were now. They were showing up when I was looking at the files changed. I will make another pass through and see if anything jumps out at me. Adding `hasImage` makes sense when people are using the unfiltered search.

I am reluctant about changing the `skip` parameter to `skipBeforeFiltering` as this makes the image search abstraction a bit leaky. Switching it back to `skip` also makes it so `nextSkip` is no longer needed and the data structure is a little simpler for users. As far as `searchForImages` (maybe `searchImageContent` would be better since we aren't getting images back but content that has images available...? or `searchContentWithImages`?), the biggest performance bottleneck will be when making the request. I don't expect larger requests to add much perf overhead (though this would be good to doublecheck). 

In other words, I suspect it would be a bit better to request larger chunks (maybe 1k at a time) and then skip and filter appropriately when requesting results with images. We could also bound the max value of `count` to something like 500 so we don't have to make so many requests for a single RPC call. 

On a smaller, related note, I had actually been imagining that you would be filtering and making the requests in `searchForImages` rather than making `_raw_search` more complex. Feel free to implement it either way, of course, but I think it was cleaner before when `_raw_search` was only responsible for searching via the API and converting the results to the expected format. I think this resulted in a nice separation of concerns where the low level function was hiding the complexity of the response format, etc, and the RPCs themselves could handle any filtering concerns.

dragazo commented:
New commit handles Smithsonian filtering skip properly (no longer skip before filter). Filtering grabs 1000 at a time (apparently any more than this and their API rejects it and instead gives 10).

I split the raw search logic apart, but for the reason that the new filter logic wouldn't be able to optimize the non-filtering skip param being the real skip param. The only reason it was together before is that it could short-circuit the parsing/construction of the object since it filtered before all that.

I also set upper bounds for the fetch counts (NYPL and Smithsonian) of 1000. For Smthsonian, this is a hard cap of their API, for NYPL it seems to be unbounded on their end (or at least >7700-ish), so it's just arbitrary. I also set the minimums of 1 that you mentioned. My only concern is that passing 0 from a variable/expr and getting an error instead of empty list might be slightly unintuitive, but I guess it's a good teaching point for the importance of testing pre-conditions.

brollb reviewed (approved):
Nice work, @dragazo! LGTM

brollb commented:
Oh, about your earlier comment:

> My only concern is that passing 0 from a variable/expr and getting an error instead of empty list might be slightly unintuitive, but I guess it's a good teaching point for the importance of testing pre-conditions.

The good news is that the types in the annotations are enforced by the netsblox server so passing `0` from the netsblox client should result in a [meaningful error](https://github.com/NetsBlox/NetsBlox/blob/b0a90295edf80ad018ff9969a01fd3ad658b5338/src/server/services/input-types.js#L36).

brollb commented:
Failing tests appear related to https://github.com/NetsBlox/NetsBlox/issues/2860 and are not due to this PR.
